File: CodeWars/Java/7Kyu/ColouredTriangles.java
// URL: https://www.codewars.com/kata/5a25ac6ac5e284cfbe000111/train/java

import java.util.*;

public class ColouredTriangles {

    public static char triangle(final String row) {
        Hashtable<String, String> combinations = new Hashtable<String, String>() {
            {
                put("GG", "G");
                put("RR", "R");
                put("BB", "B");
                put("BG", "R");
                put("GR", "B");
                put("BR", "G");
            }
        };

        List<String> colorsList = new LinkedList<String>(Arrays.asList(row.split("")));

        boolean lastCharacterFound = false;

        while (lastCharacterFound == false) {

            for (int i = 0; i < colorsList.size(); i++) {

                if ((i + 1) >= colorsList.size()) {

                    if (colorsList.size() == 1) {
                        return colorsList.get(0).charAt(0);
                    } else {
                        colorsList.remove(i);

                        if (colorsList.size() == 1) {
                            return colorsList.get(0).charAt(0);
                        }
                    }
                } else {

                    String firstColor = colorsList.get(i);

                    String secondColor = colorsList.get(i + 1);

                    String[] combination = { firstColor, secondColor };
                    Arrays.sort(combination);

                    colorsList.set(i, combinations.get(String.format("%s%s", combination[0], combination[1])));
                }
            }
        }

        return colorsList.get(0).charAt(0);
    }

    public static void main(String[] args) {
        System.out.println(triangle("RGBG"));
    }
}

File: CodeWars/JavaScript/7Kyu/ColouredTriangles.js
// URL: https://www.codewars.com/kata/5a25ac6ac5e284cfbe000111/train/javascript

function triangle(row) {
    let combinations = {"GG":"G", "RR":"R", "BB":"B", "BG":"R", "GR":"B", "BR":"G"};

    let colorsList = row.split("");
    let lastCharacterFound = false;

    while (lastCharacterFound == false){
        for (let i=0; i<colorsList.length; i++){

            if ((i+1)>=colorsList.length){
                if (colorsList.length === 1){
                    return colorsList[0];
                }
                else{
                    colorsList.pop();

                    if (colorsList.length === 1){
                        return colorsList[0];
                    }
                }
            }
            else{
                firstColor = colorsList[i];
                secondColor = colorsList[i+1];
                combination = [firstColor, secondColor].sort();
                colorsList[i] = combinations[`${combination[0]}${combination[1]}`];
            }
        }
    }
}

console.log(triangle("RGBG"));